test/integration/leds_integration_test.cc

```diff
@@ -0,0 +1,828 @@
+#include "envoy/config/bootstrap/v3/bootstrap.pb.h"
+#include "envoy/config/cluster/v3/cluster.pb.h"
+#include "envoy/config/core/v3/health_check.pb.h"
+#include "envoy/config/endpoint/v3/endpoint.pb.h"
+#include "envoy/type/v3/http.pb.h"
+
+#include "test/common/grpc/grpc_client_integration.h"
+#include "test/config/utility.h"
+#include "test/integration/http_integration.h"
+#include "test/test_common/network_utility.h"
+
+#include "gtest/gtest.h"
+
+namespace Envoy {
+namespace {
+
+// Integration test for LEDS features. CDS is consumed vi filesystem subscription,
+// and EDS and LEDS are consumed via using the delta-xDS gRPC protocol.
+class LedsIntegrationTest : public Grpc::GrpcClientIntegrationParamTest,
+                            public HttpIntegrationTest {
+protected:
+  struct FakeUpstreamInfo {
+    FakeHttpConnectionPtr connection_;
+    FakeUpstream* upstream_{};
+    absl::flat_hash_map<std::string, FakeStreamPtr> stream_by_resource_name_;
+
+    static constexpr char default_stream_name[] = "default";
+
+    // Used for cases where only a single stream is needed.
+    FakeStreamPtr& defaultStream() { return stream_by_resource_name_[default_stream_name]; }
+  };
+
+  LedsIntegrationTest()
+      : HttpIntegrationTest(Http::CodecClient::Type::HTTP1, ipVersion()),
+        codec_client_type_(envoy::type::v3::HTTP1) {
+    use_lds_ = false;
+    create_xds_upstream_ = false;
+    // LEDS is only supported by delta-xDS.
+    sotw_or_delta_ = Grpc::SotwOrDelta::Delta;
+  }
+
+  ~LedsIntegrationTest() {
+    // First disconnect upstream connections to avoid FIN messages causing unexpected
+    // disconnects on the fake servers.
+    for (auto& host_upstream_info : hosts_upstreams_info_) {
+      resetFakeUpstreamInfo(host_upstream_info);
+    }
+
+    resetFakeUpstreamInfo(leds_upstream_info_);
+    resetFakeUpstreamInfo(eds_upstream_info_);
+  }
+
+  // Printing all counters and their values
+  void printAllCounters() {
+    std::cerr << "Counters:" << std::endl;
```

Comment:
Why not use regular Envoy logging? This seems like the kind of debug we  should have in a common util.

Reply:
Sorry, this shouldn't have been pushed.
Adding this function to common testing util is possible, although it outputs many lines, and I think should only be used for development.